File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.Arrays;

class Solution {

    private int[] parent;
    private int[] rank_;

    private int find(int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    private boolean union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;

        if (rank_[ra] < rank_[rb]) {
            int tmp = ra; ra = rb; rb = tmp;
        }
        parent[rb] = ra;
        if (rank_[ra] == rank_[rb]) rank_[ra]++;
        return true;
    }

    public int solution(int n, int[][] costs) {
        Arrays.sort(costs, (x, y) -> x[2] - y[2]);

        parent = new int[n];
        rank_ = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;

        int answer = 0;
        for (int[] edge : costs) {
            int a = edge[0], b = edge[1], cost = edge[2];
            if (union(a, b)) {
                answer += cost;
            }
        }
        return answer;
    }
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
from heapq import heappush, heappop


def find_parent(parent, x):
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]


def union_parent(parent, a, b):
    a = find_parent(parent, a)
    b = find_parent(parent, b)
    
    if a < b:
        parent[b] = a
    else:
        parent[a] = b


def solution(n, costs):
    answer = 0
    q = []
    
    for a, b, c in costs:
        heappush(q, (c, a, b))
        
    parent = list(range(n))
    
    while q:
        c, a, b = heappop(q)
        if find_parent(parent, a) != find_parent(parent, b):
            union_parent(parent, a, b)
            answer += c
    
    return answer